### PR TITLE
Use AWSEKSAuthenticator in legacy path to bootstrap Ubuntu nodes

### LINKED ISSUE #3859: Ubuntu AMIs for K8s 1.20 fail to find credentials
**What were you trying to accomplish?**
Creating a new nodegroup after upgrading the cluster to 1.20. I was using an AMI based on the Ubuntu AMI for EKS version 1.20 (20210609). 

**What happened?**
The nodes failed to join the cluster. I inspected the logs of the kubelet on a node and found this.

```
2021-06-15T15:01:48Z kubelet-eks.daemon[15992]: E0615 15:01:48.247341   15992 reflector.go:138] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46: Failed to watch *v1.Pod: failed
 to list *v1.Pod: Get "https://[redacted].eu-west-1.eks.amazonaws.com/api/v1/pods?fieldSelector=spec.nodeName%3Dip-172-30-10-18.eu-west-1.compute.internal&
limit=500&resourceVersion=0": getting credentials: exec: executable heptio-authenticator-aws not found
2021-06-15T15:01:48Z kubelet-eks.daemon[15992]: It looks like you are trying to use a client-go credential plugin that is not installed.
2021-06-15T15:01:48Z kubelet-eks.daemon[15992]: To learn more about this feature, consult the documentation available at:
2021-06-15T15:01:48Z kubelet-eks.daemon[15992]:       https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins
```

**How to reproduce it?**

```
eksctl create nodegroup --config-file cluster.yaml
```

<details>
<summary>cluster.yaml</summary>

```
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: eks-prod-1
  region: eu-west-1
  version: "1.20"

vpc:
  id: vpc-***
  cidr: 172.30.0.0/16
  subnets:
    # must provide 'private' and/or 'public' subnets by availibility zone as shown
    private:
      eu-west-1a: {id: subnet-***}
      eu-west-1b: {id: subnet-***}
      eu-west-1c: {id: subnet-***}
    public:
      eu-west-1a: {id: subnet-***}
      eu-west-1b: {id: subnet-***}
      eu-west-1c: {id: subnet-***}

nodeGroups:
  - name: standard-workers-11
    minSize: 8
    maxSize: 8
    desiredCapacity: 8
    labels:
      role: workers
      lifecycle: Ec2OnDemand
    instanceType: t3a.xlarge
    ami: ami-***
    amiFamily: Ubuntu2004
    kubeletExtraConfig:
      evictionHard:
        nodefs.available: "20%"
    ssh:
      publicKeyName: production
    tags:
      Name: EksWorkers1
      Group: infra
      Project: infrastructure
      Component: k8s
      Application: k8s
      Environment: production
    preBootstrapCommands:
      - do-ansible-pull eks.yml
    securityGroups:
      attachIDs: ["sg-***"]
    volumeSize: 80
    iam:
      attachPolicyARNs:
        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
        - arn:aws:iam::aws:policy/ElasticLoadBalancingFullAccess
        # [...]
      withAddonPolicies:
        ebs: true

cloudWatch:
  clusterLogging:
    enableTypes: ["*"]

iam:
  withOIDC: true
```
</details>

**Logs**

<details>
<summary>eksctl output</summary>

```
2021-06-15 16:53:04 [ℹ]  eksctl version 0.53.0
2021-06-15 16:53:04 [ℹ]  using region eu-west-1
2021-06-15 16:53:06 [!]  Custom AMI detected for nodegroup standard-workers-11. Please refer to https://github.com/weaveworks/eksctl/issues/3563 for upcoming breaking changes
2021-06-15 16:53:06 [ℹ]  nodegroup "standard-workers-11" will use "ami-0a79ea87240daa549" [Ubuntu2004/1.20]
[...]
2021-06-15 16:53:15 [ℹ]  checking cluster stack for missing resources
2021-06-15 16:53:16 [ℹ]  cluster stack has all required resources
2021-06-15 16:53:16 [ℹ]  building nodegroup stack "eksctl-eks-prod-1-nodegroup-standard-workers-11"
2021-06-15 16:53:16 [!]  Custom AMI detected for nodegroup standard-workers-11, using legacy nodebootstrap mechanism. Please refer to https://github.com/weaveworks/eksctl/issues/3563 for upcoming breaking changes
2021-06-15 16:53:17 [ℹ]  deploying stack "eksctl-eks-prod-1-nodegroup-standard-workers-11"
2021-06-15 16:53:17 [ℹ]  waiting for CloudFormation stack "eksctl-eks-prod-1-nodegroup-standard-workers-11"
2021-06-15 16:53:35 [ℹ]  waiting for CloudFormation stack "eksctl-eks-prod-1-nodegroup-standard-workers-11"
[...]
2021-06-15 16:57:34 [ℹ]  waiting for CloudFormation stack "eksctl-eks-prod-1-nodegroup-standard-workers-11"
2021-06-15 16:57:35 [ℹ]  no tasks
2021-06-15 16:57:35 [ℹ]  adding identity "arn:aws:iam::107593380771:role/eksctl-eks-prod-1-nodegroup-stand-NodeInstanceRole-1EU5YO2L7WB7H" to auth ConfigMap
2021-06-15 16:57:35 [ℹ]  nodegroup "standard-workers-11" has 0 node(s)
2021-06-15 16:57:35 [ℹ]  waiting for at least 8 node(s) to become ready in "standard-workers-11"
Error: timed out (after 25m0s) waiting for at least 8 nodes to join the cluster and become ready in "standard-workers-11"
```
</details>

**Anything else we need to know?**

**Versions**

```
$ eksctl version
0.53.0
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:11:29Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.4-eks-6b7464", GitCommit:"6b746440c04cb81db4426842b4ae65c3f7035e53", GitTreeState:"clean", BuildDate:"2021-03-19T19:33:03Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by ilpianista**
### Description

[Heptio isn't ship anymore](https://bugs.launchpad.net/cloud-images/+bug/1932049/comments/2) in the Ubuntu 20.04 AMI. This change is needed to use `aws eks get-token` instead.

Fixes #3859

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/nodebootstrap/legacy/ubuntu.go [hunk 1]
```diff
@@ -59,7 +59,13 @@ func (b UbuntuBootstrapper) UserData() (string, error) {
 }
 
 func makeUbuntuConfig(spec *api.ClusterConfig, ng *api.NodeGroup) ([]configFile, error) {
-	clientConfigData, err := makeClientConfigData(spec, kubeconfig.HeptioAuthenticatorAWS)
+	authenticator := kubeconfig.HeptioAuthenticatorAWS
+	// Ubuntu 20.04 doesn't ship anymore heptio
+	if ng.AMIFamily == api.NodeImageFamilyUbuntu2004 {
```
Comment (toabctl) [inline pkg/nodebootstrap/legacy/ubuntu.go:64]: That's not true. The 1.20 images do no longer ship heptio. The 1.19 images (also 20.04 based) do ship heptio.

But all latest images Ubuntu ships have a recent enough `aws-cli` version so we can unconditionally switch to `AWSEKSAuthenticator`

**Review by toabctl — CHANGES_REQUESTED** [review]

File: pkg/nodebootstrap/legacy/ubuntu.go (hunk 1 above)
Comment (ilpianista) [reply pkg/nodebootstrap/legacy/ubuntu.go]: Oh. Thanks for clarifying this! I updated this PR to reflect what you said.

File: pkg/nodebootstrap/legacy/ubuntu.go [hunk 2]
```diff
@@ -59,7 +60,13 @@ func (b UbuntuBootstrapper) UserData() (string, error) {
 }
 
 func makeUbuntuConfig(spec *api.ClusterConfig, ng *api.NodeGroup) ([]configFile, error) {
-	clientConfigData, err := makeClientConfigData(spec, kubeconfig.HeptioAuthenticatorAWS)
+	supportsAWSEKS, err := utils.IsMinVersion(api.Version1_20, spec.Metadata.Version)
```
Comment (Callisto13) [inline pkg/nodebootstrap/legacy/ubuntu.go:63]:
```suggestion
	supportsAWSEKS, err := utils.IsMinVersion(api.Version1_20, spec.Metadata.Version)
	if err != nil {
		return nil, err
	}
```
needs an error check here

**Comment by Callisto13** [comment]
👍  Thanks @ilpianista, just one comment above

File: pkg/nodebootstrap/legacy/ubuntu.go [hunk 3]
```diff
@@ -59,7 +60,16 @@ func (b UbuntuBootstrapper) UserData() (string, error) {
...
```
Comment (toabctl) [inline pkg/nodebootstrap/legacy/ubuntu.go:63]: Can we please switch unconditionally to `AWSEKSAuthenticator`? The heptio version shipped in the image is old and it eg. does not work with metadata v2 only (see this bug: https://bugs.launchpad.net/cloud-images/+bug/1925487 ). That's also what the default `bootstrap.sh` script in all (all=all k8s versions) latest images do.

**Review by toabctl — CHANGES_REQUESTED** [review]

File: pkg/nodebootstrap/legacy/ubuntu.go (hunk 3 above)
Comment (ilpianista) [reply pkg/nodebootstrap/legacy/ubuntu.go:63]: And what about people using older AMIs? If we do, IMHO this should be highlighted as a breaking change maybe.

Comment (toabctl) [reply pkg/nodebootstrap/legacy/ubuntu.go:63]: I checked the oldest available ami (ami-0a5f7a061dc1fb866 , eu-central-1, from 2020-07-20) for [1.16](https://eu-central-1.console.aws.amazon.com/ec2/v2/home?region=eu-central-1#Images:visibility=public-images;name=ubuntu-eks/k8s_1.16/images/*;ownerAlias=099720109477;sort=desc:name) and that has already `awsci 1.18.69-1ubuntu0.18.04.1` which works ([minimum required version is 1.16.156](https://docs.aws.amazon.com/eks/latest/userguide/create-kubeconfig.html)). All other amis for all other k8s versions are newer. So this should be safe.

Comment (ilpianista) [reply pkg/nodebootstrap/legacy/ubuntu.go]: oh, well.

**Review by toabctl — APPROVED** [review]

**Review by cPu1 — APPROVED** [review]
@ilpianista thanks for the quick PR!

@toabctl thanks for confirming that the oldest available AMI has a recent `aws` binary that supports `get-token`.

**Pull request merged**